File: Queue/C++/QueueArray.cpp
#include <iostream>
#include "QueueArray.h"

using namespace std;

template <class T>
QueueArray<T>::QueueArray()
{
    arr = new T[10];
    this->size = 10;
    this->length = 0;
    this->front = -1;
    this->back = -1;
}

template <class T>
QueueArray<T>::QueueArray(int size){
    arr = new T[size];
    this->size =size;
    this->length = 0;
    this->front = -1;
    this->back = -1;
}

template <class T>
void QueueArray<T>::enqueue(T value){
    if(length >= size){
        cout << "Queue Overflow" << endl;
        return;
    }
    else if(front < 0){
        arr[++front] = value;
        back++;
    }
    else if(front >= size-1){
        front = -1;
        arr[++front] = value;
    }
    else{
        arr[++back] = value;
    }
    length++;
}

template <class T>
void QueueArray<T>::dequeue(){
    if(front > back || length <= 0){
        cout << "Queue Underflow" << endl;
        return;
    }
    front++;
    length--;
}

template <class T>
void QueueArray<T>::print(){
    if(front == back || length <= 0){
        cout << "Queue is empty" << endl;
        return;
    }
    for(int i=front; i<length; i++)
        cout << arr[i] << " ";
    cout << endl;
}

template <class T>
T QueueArray<T>::getFront(){
    if(length <= 0){
        cout << "Queue is empty" << endl;
        return NULL;
    }
    return arr[front];
}

template <class T>
T QueueArray<T>::getBack(){
    if(length <= 0){
        cout << "Queue is empty" << endl;
        return NULL;
    }
    return arr[back];
}

template <class T>
int QueueArray<T>::getLength(){
    return length;
}

template <class T>
int QueueArray<T>::getSize(){
    return size;
}

File: Queue/Java/QueueArray.java
public class QueueArray<T> {
    private T[] arr;
    private int front,back,length,size;
    QueueArray(){
        arr = (T[]) new Object[10];
        this.size = 10;
        this.length = 0;
        this.front = -1;
        this.back = -1;
    }
    QueueArray(int size){
        arr = (T[]) new Object[size];
        this.size = size;
        this.length = 0;
        this.front = -1;
        this.back = -1;
    }
    void enqueue(T value){
        if(length >= size){
            System.out.println("Queue Overflow");
            return;
        }
        else if(front <= -1){
            arr[++front] = value;
            back ++;
        }
        else {
            arr[++back] = value;
        }
        length++;
    }
    void dequeue(){
        if(length <= 0 || front > back){
            front = back = -1;
            System.out.println("Queue Underflow");
            return;
        }
        front--;
        length--;
    }
    void print(){
        if(length <= 0){
            System.out.println("Queue is empty");
            return;
        }
        for(int i=front; i<length; i++){
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }
    T getFront(){
        if(length <= 0){
            System.out.println("Queue is empty");
            return null;
        }
        return arr[front];
    }
    T getBack(){
        if(length <= 0){
            System.out.println("Queue is empty");
            return null;
        }
        return arr[back];
    }
    int getLength(){
        return length;
    }
    int getSize(){
        return size;
    }
}
